Include adjustments with negative taxes on Sales Tax reports

pacodelaluna opened the pull request:
#### What? Why?

Closes #5428

On Sales Tax reports, with the Tax Rates filter, you can see that the tax is not managed when an adjustment is negative. This was because we were not including negative tax rate in the calculation, so I just removed the scope applying this filter.


#### What should we test?
Add a negative adjustment on an order. Then go on `/admin/reports/sales_tax`, select the `Tax Rates` filter, then control that you can see the negative amount for the tax.


#### Release notes
- Deal with negative tax amounts for adjustments on Sales Tax reports


Changelog Category: Changed

pacodelaluna commented on app/models/spree/order_decorator.rb:
```diff
@@ -320,13 +322,6 @@ def tax_adjustment_totals
-       Spree::Money.new(tax_amount, currency: currency)]
-    end]
-  end
-
```
This method is not used anywhere.

luisramos0 commented on app/models/spree/order_decorator.rb:
```diff
@@ -305,12 +305,14 @@ def total_tax
   def tax_adjustments
-    adjustments.with_tax +
-      line_items.includes(:adjustments).map { |li| li.adjustments.with_tax }.flatten
+    adjustments + price_adjustments
```
if you want to include negatives you can change the scope to include negatives
scope :with_tax,       -> { where('spree_adjustments.included_tax > 0') }

but here you are just including all adjustments, including the ones with zero tax. that means the method become non-sense where tax_adjustments returns all adjustments.

I think you want to keep the method as tax_adjustment but exclude included_tax ZERO.

pacodelaluna replied:
Yes, you are right, I was not thinking of this case, and it will also solve the comment you put above for `adjustment.included_tax.zero?` control. I will fix this, need to cross-check side effect related to this scope use.

pacodelaluna commented:
@luisramos0 Thanks for your feedback, I will take a look at it this weekend, extract the logic into a service makes sense yes.

pacodelaluna commented:
@luisramos0 Sorry I was busy lately but I have refactored the code within a service. I have changed the scope to now include also the negative tax values, there will be a side effect but I think it is a desired effect. Tell me if any issue.

sauloperez commented:
it doesn't look good. Using `PROFILE` from master:

![Screenshot from 2020-06-11 12-12-42](https://user-images.githubusercontent.com/762088/84373572-0d733e80-abdd-11ea-817f-68700c4823b0.png)

Using `PROFILE` from this branch:

![Screenshot from 2020-06-11 12-13-14](https://user-images.githubusercontent.com/762088/84373623-20860e80-abdd-11ea-8d86-a38c473d65bb.png)

pacodelaluna commented:
@sauloperez Oh ok, I am a bit surprised about this performance issue as I thought I was mostly moving code inside the PR. I will take a deeper look thsi weekend, I saw that there was an other service, ` OrderAdjustmentsFetcher`, I initially didn't want to add too much refactoring but it seems we really need it...

sauloperez commented:
I'm pretty sure it's scoped within `OrderTaxAdjustmentsFetcher#all`. Surely can be refactored to a single query starting from `Spree:Adjustment`.

sauloperez commented:
@pacodelaluna I gave it a try and added a couple of commits. Take a look at them and let me know what you think. They remove the N+1 and effectively fetch the adjustments in one shot.

There is at least one more N+1 related to adjustments in this report but this PR brings already a better world.

pacodelaluna commented:
@sauloperez Great job! Thanks for your help on this!

sigmundpetersen commented:
The conflict on this branch needs to be resolved before testing. A rebase would be good?

pacodelaluna commented:
@sigmundpetersen Just did the rebase, should be fine for testing.

filipefurtad0 commented:
Hey @pacodelaluna!

I checked/reproduced what @RachL described in the issue. So, placed order R006812641, which can be seen in Tax Rates report:

![image](https://user-images.githubusercontent.com/49817236/84810625-828eab80-b003-11ea-8d4b-d9373d498db6.png)

Below, the same order, after adjustment with the amount £-20.00 (£-3.33 included tax):
![image](https://user-images.githubusercontent.com/49817236/84810939-fc269980-b003-11ea-89d5-04bbffec5ebe.png)


After this PR the report is rendered with the correct values:
![image](https://user-images.githubusercontent.com/49817236/84812118-ce425480-b005-11ea-9dbb-b2682cb67a8a.png)

Awesome, thanks Paco :tada: 
Ready to go.

PS: I found that clicking on that order-hyperlink leads to /admin/orders/R006812641 -> which is neither related to this PR nor to the upgrade - checked on staging-es as well. Opening a separate issue.
